### PR TITLE
feat(dev): Stop code generation action

### PR DESCRIPTION
**Pull request by tverney**
## Problem

Currently /dev don't support stop code generation. This PR introduces this functionality (watch the video below).

## Solution

Mynah UI provides onStopChatResponse API which we can hook in the cancellation token provided on VS Code, sharing across an active session, aborting current progress. 

https://github.com/user-attachments/assets/7a563bdc-c26b-4052-92c5-df0eeed63ea8

Notes:

- RTS is already supporting the change on prod
- codeGenerationId and uploadId are now being generated in FE
- AppSec and pentest passed successfully 
- Telemetry is included in this PR and strings are reviewed/approved by Technical writer
- This stop action is only introduced in /dev feature. This did not enable in any other product for this extension.

---

License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by tverney** [comment]
!!Important - FYI toolkit team: this is only to be merged in next week's release. CC: @neilk-aws

**Review by neilk-aws — CHANGES_REQUESTED** [review]
Blocking to wait for security sign-off

**Comment by tverney** [comment]
We have already the security sign-off. Ready to be reviewed cc: @neilk-aws

File: packages/core/src/amazonq/webview/ui/connector.ts
Comment (jpinkney-aws) [inline packages/core/src/amazonq/webview/ui/connector.ts:465]: Does this break anything on cwc's side?

File: packages/core/src/shared/telemetry/vscodeTelemetry.json
Comment (jpinkney-aws) [inline packages/core/src/shared/telemetry/vscodeTelemetry.json:1200]: What information will having a tabID give us? Isn't it just a random string like xcfg2? Instead does it make sense to use the telemetry.ui_click metric instead?

File: packages/core/src/amazonq/webview/ui/followUps/handler.ts
Comment (jpinkney-aws) [inline packages/core/src/amazonq/webview/ui/followUps/handler.ts:49]: Out of curiosity does this mean that every single update store will need this? Why isn't the default false and then its true when we want the cancel button (which I assume is the stop button) to show?

Comment (jpinkney-aws) [reply packages/core/src/amazonq/webview/ui/followUps/handler.ts:49]: The change itself kind of scares me because we need to make sure its propagated everything _and_ that everyone who makes changes to the updateStore knows why it exists

File: packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts
Comment (jpinkney-aws) [inline packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts:316]: Can we add a comment on this function why its needed? I assume its because we want to refresh the token on every request? Or in this approach we only dispose it when we cancel something -- since otherwise the token will still be valid?

File: packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts
Comment (jpinkney-aws) [inline packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts:502]: If remainingIterations is 0 when undefined can we just add that into the function definition `remainingIterations: number = 0`

File: packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts
Comment (jpinkney-aws) [inline packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts:782]: I think I mentioned this in a different file but is there any reason we can't just use `telemetry.ui_click` here? since I believe this only being called when a user clicks on the stop generating button and tabID doesn't really give us anything valuable

File: packages/core/src/amazonqFeatureDev/session/sessionState.ts
Comment (jpinkney-aws) [inline packages/core/src/amazonqFeatureDev/session/sessionState.ts:47]: is this id some special field on the backend?

File: packages/core/src/amazonqFeatureDev/session/sessionState.ts
Comment (jpinkney-aws) [inline packages/core/src/amazonqFeatureDev/session/sessionState.ts:303]: Why are lies 292-297 and this the exact same?

File: packages/core/src/amazonqFeatureDev/session/sessionState.ts
Comment (jpinkney-aws) [inline packages/core/src/amazonqFeatureDev/session/sessionState.ts:499]: what does superTokenSource mean? also instead of falling back to a default you can just do:
```
superTokenSource: vscode.CancellationTokenSource = new vscode.CancellationTokenSource()
```

File: packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts
Comment (tverney) [reply packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts:502]: Good call, I'll update

File: packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts
Comment (tverney) [reply packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts:782]: This change came from @kelvin-klchu, adding him here in the thread if makes sense or not.

File: packages/core/src/shared/telemetry/vscodeTelemetry.json
Comment (tverney) [reply packages/core/src/shared/telemetry/vscodeTelemetry.json:1200]: I'm pulling @kelvin-klchu to answer these

File: packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts
Comment (tverney) [reply packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts:316]: Sure, I'll provide a more clearer comment there. Essentially when user stops he still in the same session, so we need clean up the token and create a new one for user provide a new prompt.

File: packages/core/src/amazonqFeatureDev/session/sessionState.ts
Comment (tverney) [reply packages/core/src/amazonqFeatureDev/session/sessionState.ts:303]: Good call, rebased / merge issue. Just fixing this one

File: packages/core/src/amazonq/webview/ui/connector.ts
Comment (tverney) [reply packages/core/src/amazonq/webview/ui/connector.ts:465]: No, cwc don't use it currently.

File: packages/core/src/amazonq/webview/ui/followUps/handler.ts
Comment (dogusata) [reply packages/core/src/amazonq/webview/ui/followUps/handler.ts:49]: If none of the functionalities are supporting `onStopChatResponse` why we have it attached to the mynahui instance? If some functions are using it, then this change makes sense. However the defaults can be changed for newly generated tabs. But i don't see any specific difference between giving true to the tabs which will using it or false to tabs which will not.

Comment (dogusata) [reply packages/core/src/amazonq/webview/ui/followUps/handler.ts:49]: Additionally, giving the value during the tab generation is sufficient. Also, until send a new value, the previous value will remain.

Comment (tverney) [reply packages/core/src/amazonq/webview/ui/followUps/handler.ts:49]: @jpinkney-aws do we have a consensus about this specific ? Or do we need sync about it? I'm ok for now with @dogusata, no one its using it, I did safe check in all the places in this PR that we use the store.

Comment (hayemaxi) [reply packages/core/src/amazonq/webview/ui/followUps/handler.ts:49]: So just to confirm - state of `cancelButtonWhenLoading = false` must be passed if `onStopChatResponse` was specified earlier, otherwise it defaults to true. If `onStopChatResponse` wasn't passed, then it doesn't matter.

If this is correct then the current implementation is fine IMO.

Comment (tverney) [reply packages/core/src/amazonq/webview/ui/followUps/handler.ts:49]: Exactly, that's the behaviour. If onStopChatResponse isn't passed, it defaults to false.

File: packages/core/src/amazonq/webview/ui/main.ts
Comment (dhasani23) [inline packages/core/src/amazonq/webview/ui/main.ts:200]: This is causing the "stop" button to always appear which is a problem for `/transform` because we have our own stop button on the card in the chat message.

Comment (dhasani23) [reply packages/core/src/amazonq/webview/ui/main.ts:200]: image

**Pull request merged**
